dhall-docs/src/Dhall/Docs/Comment.hs

```diff
@@ -0,0 +1,305 @@
+{-| Provides utilities to parse Dhall comments
+-}
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+module Dhall.Docs.Comment
+    ( parseComments
+    , CommentParseError(..)
+    , DhallDocsText
+    , parseSingleDhallDocsComment
+    , unDhallDocsText
+    ) where
+
+import Control.Applicative (many, some, (<|>))
+import Data.Functor        (void)
+import Data.List.NonEmpty  (NonEmpty (..), (<|))
+import Data.Text           (Text)
+import Dhall.Docs.Util
+import Dhall.Parser        (Parser (..))
+import Text.Megaparsec     (SourcePos, (<?>))
+
+import qualified Data.Either
+import qualified Data.Foldable
+import qualified Data.List.NonEmpty       as NonEmpty
+import qualified Data.Maybe               as Maybe
+import qualified Data.Text
+import qualified Dhall.Parser.Combinators as Combinators
+import qualified Dhall.Parser.Expression  as Expression
+import qualified Dhall.Parser.Token       as Token
+import qualified Text.Megaparsec
+import qualified Text.Megaparsec.Pos      as Megaparsec.Pos
+
+-- | For explanation of these data-types see 'DhallComment'
+data DhallDocsComment
+data MarkedComment
+data RawComment
+
+type ListOfSingleLineComments = NonEmpty (SourcePos, Text)
+
+{-| Internal representation of Dhall comments. Text is always stripped from
+    whitespace from both the start and begin of the string
+
+    * If @a = 'DhallDocsComment'@ then comment is valid to extract its contents
+    and be rendered on documentation
+    * If @a = 'MarkedComment'@ then comment has the @|@ marker that
+    @dhall-docs@ will be aware of, but this comment may not be a @dhall-docs@
+    comment
+    * If @a = `RawComment`@ then the comment is a raw comment
+-}
+data DhallComment a
+    -- | A single block comment: starting from @{-@ and ending in @-}@
+    = BlockComment Text
+    {-| A group of subsequent single line comment, each one starting from @--@
+        and ending in the last character to the end of the line. Each one keeps
+        its 'SourcePos' to validate indentation.
+
+        A property of 'SingleLineComments' is that the 'sourceLine's in the
+        'NonEmpty (SourcePos, Text)' are in strictly-increasing order /and/
+        the difference between the 'sourceLine' of any adyacent pair is @1@.
+
+        Note that several @dhall-docs@ comments maybe inside a single 'SingleLineComments'
+    -}
+    | SingleLineComments ListOfSingleLineComments
+    deriving Show
+
+-- | Extracted text from a valid @dhall-docs@ comment
+newtype DhallDocsText = DhallDocsText Text
+    deriving Show
+
+unDhallDocsText :: DhallDocsText -> Text
+unDhallDocsText (DhallDocsText t) = t
+
+-- | A mirror of "Dhall.Parser.Token".'Dhall.Parser.Token.lineComment' but
+--   returning a 'DhallComment'
+lineCommentParser :: Parser (NonEmpty (DhallComment RawComment))
```

Comment:
Would it be possible to upstream this into the `dhall` package?  That way we don't have to worry about the two implementations of this diverging

Reply:
I believe we could do it. Would you mind if I do that in a follow-up PR?

Reply:
OTOH: should we move only the parsing logic returning `Text` and keeping all data-types here? I say this because almost all defined data-types here are specific for dhall-docs comments validation

Reply:
Yeah, it's fine to do that in a follow-up PR.  I think this can be done in such a way that it does not depend on `dhall-docs`-specific types